_topics/en/merkle-tree-vulnerabilities.md

```diff
@@ -0,0 +1,268 @@
+  of the merkle tree used by the Bitcoin consensus protocol.
+
+---
+
+A merkle tree whose number of elements is a power of two can trivially
+be compressed into a merkle root: for a given set of elements (e.g. {A,
+B, C, D}) create pairs of elements (e.g. {A,B}) and hash them together;
+take the resultant internal elements (e.g., {AB, CD}}), create pairs,
+and hash them together again.  Repeat until only one element remains,
+the merkle root (e.g. {ABCD}).
+
+```text
+     ABCD
+    /    \
+  AB      CD
+ /  \    /  \
+A    B  C    D
+```
+
+Merkle trees are used to allow a single 32-byte hash in each block's
+header to commit to all of the transactions in that block in a way that
+allows creating very space-efficient proofs of inclusion (called
+_partial merkle proofs_).  However, Bitcoin blocks may contain a number
+of transactions that is not a power of two, requiring a more complicated
+algorithm to construct the merkle root.  The algorithm chosen for
+Bitcoin hashes any unbalanced element with itself:
+
+```text
+     ABCC
+    /    \
+  AB      CC
+ /  \    /
+A    B  C
+```
+
+There's also the special case of a block with only one transaction
+having its hash used as the merkle root:
+
+```text
+A
+|
+A
+```
+
+The subsequent sections describe several vulnerabilities associated with
+this algorithm, along with their deployed mitigations.
+
+## Invalidity caching
+
+Bitcoin Core prevents wasting bandwidth, CPU, and memory
+downloading and validating invalid blocks multiple times by caching the
+header hash of recently seen blocks that are known to be invalid.
+However, an attacker could modify a valid block to make it look like an
+invalid block.  Consider the following valid block from our description
+above:
+
+```text
+     ABCC
+    /    \
+  AB      CC
+ /  \    /
+A    B  C
+```
+
+The attacker can modify that block by duplicating the final transaction
+to create a block with an identical merkle root but which is invalid (as
+the second _C_ transaction spends a UTXO that was already spent in the
+first _C_ transaction, the double spending of UTXOs being forbidden):
+
+```text
+     ABCC
+    /    \
+  AB      CC
+ /  \    /  \
+A    B  C    C{% comment %}<!-- skip-test -->{% endcomment %}
+```
+
+If Bitcoin Core saw the invalid form of the block first, it would store
+its block header hash in its invalidity cache and refuse to accept the
+valid form of the block from any of its peers until the cache was
+cleared (which occurred when the node was restarted).  An attacker could
+use this to [eclipse attack][topic eclipse attacks] particular nodes.
+
+The vulnerability was [reported][voight merkle] in 2012, assigned
+[CVE-2012-2459][topic cves], and quietly fixed without ever being
+exploited (to the best of our knowledge).
+
+A [variation][news37 merkle] of the vulnerability was re-introduced into
+Bitcoin Core in version 0.13.0 (2016) and quietly fixed in version
+0.14.0 (2017).  Another [variation][news251 merkle] was discovered and
```

Comment:
This 251 link isnt clear to me, is the PR the fix?

(edit: ok I get it now after the mention "Bitcoin Core #27608 protects against an invalidity caching vulnerability" connected the dots for me)

Reply:
Moved the link to the verb "fixed" to make it clear.